File: educational-core/src/com/jetbrains/edu/coursecreator/actions/placeholder/CCAddAnswerPlaceholder.java
package com.jetbrains.edu.coursecreator.actions.placeholder;

import com.intellij.openapi.actionSystem.Presentation;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.SelectionModel;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.project.Project;
import com.intellij.util.DocumentUtil;
import com.jetbrains.edu.coursecreator.CCUtils;
import com.jetbrains.edu.coursecreator.actions.placeholder.CCCreateAnswerPlaceholderDialog.DependencyInfo;
import com.jetbrains.edu.learning.EduState;
import com.jetbrains.edu.learning.EduUtils;
import com.jetbrains.edu.learning.PlaceholderPainter;
import com.jetbrains.edu.learning.StudyTaskManager;
import com.jetbrains.edu.learning.configuration.EduConfigurator;
import com.jetbrains.edu.learning.courseFormat.AnswerPlaceholder;
import com.jetbrains.edu.learning.courseFormat.AnswerPlaceholderDependency;
import com.jetbrains.edu.learning.courseFormat.Course;
import com.jetbrains.edu.learning.courseFormat.TaskFile;
import com.jetbrains.edu.learning.courseFormat.ext.CourseExt;
import com.jetbrains.edu.learning.messages.EduCoreBundle;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class CCAddAnswerPlaceholder extends CCAnswerPlaceholderAction {

  public CCAddAnswerPlaceholder() {
    super(EduCoreBundle.lazyMessage("action.add.answer.placeholder.text"),
          EduCoreBundle.lazyMessage("action.add.answer.placeholder.description"));
  }


  private static boolean arePlaceholdersIntersect(@NotNull final TaskFile taskFile, int start, int end) {
    List<AnswerPlaceholder> answerPlaceholders = taskFile.getAnswerPlaceholders();
    for (AnswerPlaceholder existingAnswerPlaceholder : answerPlaceholders) {
      int twStart = existingAnswerPlaceholder.getOffset();
      int twEnd = existingAnswerPlaceholder.getPossibleAnswer().length() + twStart;
      if ((start >= twStart && start < twEnd) || (end > twStart && end <= twEnd) ||
          (twStart >= start && twStart < end) || (twEnd > start && twEnd <= end)) {
        return true;
      }
    }
    return false;
  }

  private void addPlaceholder(@NotNull Project project, @NotNull EduState state) {
    Editor editor = state.getEditor();
    Document document = editor.getDocument();
    FileDocumentManager.getInstance().saveDocument(document);
    final SelectionModel model = editor.getSelectionModel();
    final int offset = model.hasSelection() ? model.getSelectionStart() : editor.getCaretModel().getOffset();
    TaskFile taskFile = state.getTaskFile();
    final AnswerPlaceholder answerPlaceholder = new AnswerPlaceholder();
    int index = taskFile.getAnswerPlaceholders().size();
    answerPlaceholder.setIndex(index);
    answerPlaceholder.setTaskFile(taskFile);
    taskFile.sortAnswerPlaceholders();
    answerPlaceholder.setOffset(offset);

    @NonNls String defaultPlaceholderText = getDefaultPlaceholderText(project);
    answerPlaceholder.setPlaceholderText(defaultPlaceholderText);
    CCCreateAnswerPlaceholderDialog dlg = createDialog(project, answerPlaceholder);
    if (!dlg.showAndGet()) {
      return;
    }
    String answerPlaceholderText = dlg.getPlaceholderText();
    String possibleAnswer = model.hasSelection() ? model.getSelectedText() : defaultPlaceholderText;
    if (possibleAnswer == null) {
      possibleAnswer = defaultPlaceholderText;
    }
    answerPlaceholder.setPlaceholderText(answerPlaceholderText);
    answerPlaceholder.setLength(possibleAnswer.length());
    final DependencyInfo dependencyInfo = dlg.getDependencyInfo();
    if (dependencyInfo != null) {
      answerPlaceholder.setPlaceholderDependency(
        AnswerPlaceholderDependency.create(answerPlaceholder, dependencyInfo.getDependencyPath(), dependencyInfo.isVisible()));
    }

    if (!model.hasSelection()) {
      DocumentUtil.writeInRunUndoTransparentAction(() -> document.insertString(offset, defaultPlaceholderText));
    }

    AddAction action = new AddAction(project, answerPlaceholder, taskFile, editor);
    EduUtils.runUndoableAction(project, EduCoreBundle.message("action.add.answer.placeholder.text"), action);
  }

  @NotNull
  private static String getDefaultPlaceholderText(@NotNull final Project project) {
    final Course course = StudyTaskManager.getInstance(project).getCourse();
    if (course == null) {
      return CCUtils.DEFAULT_PLACEHOLDER_TEXT;
    }
    EduConfigurator<?> configurator = CourseExt.getConfigurator(course);
    if (configurator == null) {
      return CCUtils.DEFAULT_PLACEHOLDER_TEXT;
    }
    return configurator.getDefaultPlaceholderText();
  }

  static class AddAction extends TaskFileUndoableAction {
    private final AnswerPlaceholder myPlaceholder;

    public AddAction(@NotNull Project project, @NotNull AnswerPlaceholder placeholder,
                     @NotNull TaskFile taskFile, @NotNull Editor editor) {
      super(project, taskFile, editor);
      myPlaceholder = placeholder;
    }

    @Override
    public boolean performUndo() {
      final List<AnswerPlaceholder> answerPlaceholders = getTaskFile().getAnswerPlaceholders();
      if (answerPlaceholders.contains(myPlaceholder)) {
        answerPlaceholders.remove(myPlaceholder);
        PlaceholderPainter.hidePlaceholder(myPlaceholder);
        return true;
      }

      return false;
    }

    @Override
    public void performRedo() {
      getTaskFile().addAnswerPlaceholder(myPlaceholder);
      PlaceholderPainter.showPlaceholder(getProject(), myPlaceholder);
    }
  }

  @Override
  protected void performAnswerPlaceholderAction(@NotNull Project project, @NotNull EduState state) {
    addPlaceholder(project, state);
  }

  @Override
  protected void updatePresentation(@NotNull EduState eduState, @NotNull Presentation presentation) {
    presentation.setVisible(true);
    if (canAddPlaceholder(eduState)) {
      presentation.setEnabled(true);
    }
  }

  private static boolean canAddPlaceholder(@NotNull EduState state) {
    Editor editor = state.getEditor();
    SelectionModel selectionModel = editor.getSelectionModel();
    TaskFile taskFile = state.getTaskFile();
    if (selectionModel.hasSelection()) {
      int start = selectionModel.getSelectionStart();
      int end = selectionModel.getSelectionEnd();
      return !arePlaceholdersIntersect(taskFile, start, end);
    }
    int offset = editor.getCaretModel().getOffset();
    return taskFile.getAnswerPlaceholder(offset) == null;
  }

  protected CCCreateAnswerPlaceholderDialog createDialog(Project project, AnswerPlaceholder answerPlaceholder) {
    return new CCCreateAnswerPlaceholderDialog(project, false, answerPlaceholder);
  }
}

File: intellij-plugin/educational-core/src/com/jetbrains/edu/coursecreator/actions/placeholder/CCAddAnswerPlaceholder.kt
package com.jetbrains.edu.coursecreator.actions.placeholder

import com.intellij.openapi.actionSystem.Presentation
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.project.Project
import com.intellij.util.DocumentUtil
import com.jetbrains.edu.coursecreator.CCUtils
import com.jetbrains.edu.learning.EduState
import com.jetbrains.edu.learning.StudyTaskManager
import com.jetbrains.edu.learning.actions.EduActionUtils.runUndoableAction
import com.jetbrains.edu.learning.courseFormat.AnswerPlaceholder
import com.jetbrains.edu.learning.courseFormat.AnswerPlaceholderDependency.Companion.create
import com.jetbrains.edu.learning.courseFormat.TaskFile
import com.jetbrains.edu.learning.courseFormat.ext.configurator
import com.jetbrains.edu.learning.messages.EduCoreBundle.message
import com.jetbrains.edu.learning.placeholder.PlaceholderHighlightingManager

open class CCAddAnswerPlaceholder : CCAnswerPlaceholderAction() {
  private fun addPlaceholder(state: EduState) {
    val editor = state.editor
    FileDocumentManager.getInstance().saveDocument(editor.document)
    val model = editor.selectionModel
    val offset = if (model.hasSelection()) model.selectionStart else editor.caretModel.offset
    val taskFile = state.taskFile
    val defaultPlaceholderText = defaultPlaceholderText(state.project)

    val answerPlaceholder = AnswerPlaceholder().apply {
      index = taskFile.answerPlaceholders.size
      this.taskFile = taskFile
      this.offset = offset
      placeholderText = defaultPlaceholderText
    }

    val dlg = createDialog(state.project, answerPlaceholder)
    if (!dlg.showAndGet()) {
      return
    }
    val answerPlaceholderText = dlg.getPlaceholderText()
    val selectedText = model.selectedText
    val possibleAnswer = if (model.hasSelection() && selectedText != null) selectedText else defaultPlaceholderText
    answerPlaceholder.placeholderText = answerPlaceholderText
    answerPlaceholder.length = possibleAnswer.length
    val dependencyInfo = dlg.getDependencyInfo()
    if (dependencyInfo != null) {
      answerPlaceholder.placeholderDependency = create(answerPlaceholder, dependencyInfo.dependencyPath)
    }
    answerPlaceholder.isVisible = dlg.getVisible()
    answerPlaceholder.init()
    if (!model.hasSelection()) {
      DocumentUtil.writeInRunUndoTransparentAction { editor.document.insertString(offset, defaultPlaceholderText) }
    }
    val action = AddAction(state.project, answerPlaceholder, taskFile, editor)
    runUndoableAction(state.project, message("action.Educational.Educator.AddAnswerPlaceholder.text"), action)
  }

  open fun createDialog(project: Project, answerPlaceholder: AnswerPlaceholder): CCCreateAnswerPlaceholderDialog {
    return CCCreateAnswerPlaceholderDialog(project, false, answerPlaceholder)
  }

  internal open class AddAction(
    project: Project,
    private val placeholder: AnswerPlaceholder,
    taskFile: TaskFile,
    editor: Editor
  ) : TaskFileUndoableAction(project, taskFile, editor) {
    override fun performUndo(): Boolean {
      if (taskFile.answerPlaceholders.contains(placeholder)) {
        taskFile.removeAnswerPlaceholder(placeholder)
        taskFile.sortAnswerPlaceholders()
        PlaceholderHighlightingManager.hidePlaceholder(project, placeholder)
        return true
      }
      return false
    }

    override fun performRedo() {
      taskFile.addAnswerPlaceholder(placeholder)
      taskFile.sortAnswerPlaceholders()
      PlaceholderHighlightingManager.showPlaceholder(project, placeholder)
    }
  }

  override fun performAnswerPlaceholderAction(state: EduState) {
    addPlaceholder(state)
  }

  override fun updatePresentation(eduState: EduState, presentation: Presentation) {
    presentation.isVisible = true
    if (canAddPlaceholder(eduState)) {
      presentation.isEnabled = true
    }
  }

  companion object {
    private fun arePlaceholdersIntersect(taskFile: TaskFile, start: Int, end: Int): Boolean {
      val answerPlaceholders = taskFile.answerPlaceholders
      for (existingAnswerPlaceholder in answerPlaceholders) {
        val twStart = existingAnswerPlaceholder.offset
        val twEnd = existingAnswerPlaceholder.possibleAnswer.length + twStart
        if (start in twStart until twEnd || end in (twStart + 1)..twEnd ||
            twStart in start until end || twEnd in (start + 1)..end) {
          return true
        }
      }
      return false
    }

    private fun defaultPlaceholderText(project: Project): String {
      val course = StudyTaskManager.getInstance(project).course ?: return CCUtils.DEFAULT_PLACEHOLDER_TEXT
      val configurator = course.configurator ?: return CCUtils.DEFAULT_PLACEHOLDER_TEXT
      return configurator.defaultPlaceholderText
    }

    private fun canAddPlaceholder(state: EduState): Boolean {
      val editor = state.editor
      val selectionModel = editor.selectionModel
      val taskFile = state.taskFile
      if (selectionModel.hasSelection()) {
        val start = selectionModel.selectionStart
        val end = selectionModel.selectionEnd
        return !arePlaceholdersIntersect(taskFile, start, end)
      }
      val offset = editor.caretModel.offset
      return taskFile.getAnswerPlaceholder(offset) == null
    }
  }
}
